lib/src/authentication/client.dart

```diff
@@ -34,17 +40,68 @@ class _AuthenticatedClient extends http.BaseClient {
     // to given serverBaseUrl. Otherwise credential leaks might ocurr when
     // archive_url hosted on 3rd party server that should not receive
     // credentials of the first party.
-    if (credential.canAuthenticate(request.url.toString())) {
+    if (credential != null &&
+        credential.canAuthenticate(request.url.toString())) {
       request.headers[HttpHeaders.authorizationHeader] =
           await credential.getAuthorizationHeaderValue();
     }
-    return _inner.send(request);
+
+    try {
+      final response = await _inner.send(request);
+      if (response.statusCode == 401) _handleError(response);
+      return response;
+    } on PubHttpException catch (error) {
+      if (error.response?.statusCode == 403) _handleError(error.response);
+      rethrow;
+    }
+  }
+
+  void _handleError(http.BaseResponse response) {
+    String serverMessage;
+    try {
+      final wwwAuthenticateHeaderValue =
+          response.headers[HttpHeaders.wwwAuthenticateHeader];
+      if (wwwAuthenticateHeaderValue != null) {
+        final parsedValue = HeaderValue.parse(wwwAuthenticateHeaderValue,
+            parameterSeparator: ',');
+        if (parsedValue.parameters['realm'] == 'pub') {
+          serverMessage = parsedValue.parameters['message'];
+        }
```

Comment:
I doubt this parsing complies with [RFC 7235 section 4.1](https://datatracker.ietf.org/doc/html/rfc7235#section-4.1).

Also make sure to reference `https://datatracker.ietf.org/doc/html/rfc7235#section-4.1` in code comments.
So that we clearly state that we intend to implement this specification correctly.

Problems:
1.  We never check the `auth-scheme` from the `challenge` (we only support `Bearer`).
2.  We do not support multiple `challenge` entries in `www-authenticate` header.

Relevant grammar from RFC 7235:
```
WWW-Authenticate = 1#challenge

challenge   = auth-scheme [ 1*SP ( token68 / #auth-param ) ]

auth-scheme    = token

auth-param     = token BWS "=" BWS ( token / quoted-string )

token68        = 1*( ALPHA / DIGIT /
                        "-" / "." / "_" / "~" / "+" / "/" ) *"="
```

(Notice that [section 1.2](https://datatracker.ietf.org/doc/html/rfc7235#section-1.2) specifies `#` operator as comma separated list, e.g. `#foo` is zero or more `foo` separated by commas.

**Example**
```
WWW-Authenticate: Bearer realm="apps", type=1, title="hello \"user\"", Basic realm="webdav", d2d2d2__, Bearer realm=pub, message="Get a token at \"https://other-pub.com\""
```

We might want to create a file with a function for parsing this, and write a handful of unit-tests for this.

Parsing this might not be entirely trivial as there can be commas inside the quoted strings. And quoted-string probably has a few corner cases around escaping.

In practice I doubt many servers will support multiple authentication challenges. But if we can parse these challenges correctly now, that would be fantastic. Because maybe in the future we'll see things like:

```
WWW-Authenticate: Bearer realm=pub, message="Get a token at \"https://other-pub.com\"",
                  Bearer realm=pub, serviceAccount="<service-account-name>@<project-id>.iam.gserviceaccount.com"
```

Hmm, we should probably make sure to iterate through all the challenges provided by the server! And we should probably ignore challenges that with `realm=pub` if they don't include a `message`.